## E-commerce Marketing Causal Inference & A/B Testing Project

In [1]:
import pandas as pd
from pathlib import Path

RAW = Path('/Users/dhritisood/Olist ecommerce/data/raw')

dataset_paths = {
    'olist_customers_dataset': RAW / 'olist_customers_dataset.csv',
    'olist_geolocation_dataset': RAW / 'olist_geolocation_dataset.csv',
    'olist_order_items_dataset': RAW / 'olist_order_items_dataset.csv',
    'olist_order_payments_dataset': RAW / 'olist_order_payments_dataset.csv',
    'olist_order_reviews_dataset': RAW / 'olist_order_reviews_dataset.csv',
    'olist_orders_dataset': RAW / 'olist_orders_dataset.csv',
    'olist_products_dataset': RAW / 'olist_products_dataset.csv',
    'olist_sellers_dataset': RAW / 'olist_sellers_dataset.csv',
    'product_category_name_translation': RAW / 'product_category_name_translation.csv',
}

dataframes = {}
for name, path in dataset_paths.items():
    dataframes[name] = pd.read_csv(path)
    print(f'Loaded {name} with {dataframes[name].shape[0]} rows and {dataframes[name].shape[1]} columns.')


Loaded olist_customers_dataset with 99441 rows and 5 columns.


Loaded olist_geolocation_dataset with 1000163 rows and 5 columns.
Loaded olist_order_items_dataset with 112650 rows and 7 columns.
Loaded olist_order_payments_dataset with 103886 rows and 5 columns.


Loaded olist_order_reviews_dataset with 99224 rows and 7 columns.
Loaded olist_orders_dataset with 99441 rows and 8 columns.
Loaded olist_products_dataset with 32951 rows and 9 columns.
Loaded olist_sellers_dataset with 3095 rows and 4 columns.
Loaded product_category_name_translation with 71 rows and 2 columns.


Now that the datasets are loaded, let's display the first few rows and basic information for each to understand their structure and content.

In [2]:
for name, df in dataframes.items():
    print(f'\n--- {name} ---\n')
    display(df.head())
    print(f'Info for {name}:')
    df.info()



--- olist_customers_dataset ---



,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP


Info for olist_customers_dataset:
<class 'pandas.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype
---  ------                    --------------  -----
 0   customer_id               99441 non-null  str  
 1   customer_unique_id        99441 non-null  str  
 2   customer_zip_code_prefix  99441 non-null  int64
 3   customer_city             99441 non-null  str  
 4   customer_state            99441 non-null  str  
dtypes: int64(1), str(4)
memory usage: 3.8 MB

--- olist_geolocation_dataset ---



,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,1037,-23.545621,-46.639292,sao paulo,SP
1,1046,-23.546081,-46.644820,sao paulo,SP
2,1046,-23.546129,-46.642951,sao paulo,SP
3,1041,-23.544392,-46.639499,sao paulo,SP
4,1035,-23.541578,-46.641607,sao paulo,SP


Info for olist_geolocation_dataset:


<class 'pandas.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   geolocation_zip_code_prefix  1000163 non-null  int64  
 1   geolocation_lat              1000163 non-null  float64
 2   geolocation_lng              1000163 non-null  float64
 3   geolocation_city             1000163 non-null  str    
 4   geolocation_state            1000163 non-null  str    
dtypes: float64(2), int64(1), str(2)
memory usage: 38.2 MB

--- olist_order_items_dataset ---



,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14


Info for olist_order_items_dataset:
<class 'pandas.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  str    
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  str    
 3   seller_id            112650 non-null  str    
 4   shipping_limit_date  112650 non-null  str    
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), str(4)
memory usage: 6.0 MB

--- olist_order_payments_dataset ---



,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39
2,25e8ea4e93396b6fa0d3dd708e76c1bd,1,credit_card,1,65.71
3,ba78997921bbcdc1373bb41e913ab953,1,credit_card,8,107.78
4,42fdf880ba16b47b59251dd489d4441a,1,credit_card,2,128.45


Info for olist_order_payments_dataset:
<class 'pandas.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  str    
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  str    
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), str(2)
memory usage: 4.0 MB

--- olist_order_reviews_dataset ---



,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,NaN,NaN,2018-01-18 00:00:00,2018-01-18 21:46:59
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,NaN,NaN,2018-03-10 00:00:00,2018-03-11 03:05:13
2,228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,NaN,NaN,2018-02-17 00:00:00,2018-02-18 14:36:24
3,e64fb393e7b32834bb789ff8bb30750e,658677c97b385a9be170737859d3511b,5,NaN,Recebi bem antes do prazo estipulado.,2017-04-21 00:00:00,2017-04-21 22:02:06
4,f7c4243c7fe1938f181bec41a392bdeb,8e6bfb81e283fa7e4f11123a3fb894f1,5,NaN,Parabéns lojas lannister adorei comprar pela I...,2018-03-01 00:00:00,2018-03-02 10:26:53


Info for olist_order_reviews_dataset:
<class 'pandas.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype
---  ------                   --------------  -----
 0   review_id                99224 non-null  str  
 1   order_id                 99224 non-null  str  
 2   review_score             99224 non-null  int64
 3   review_comment_title     11568 non-null  str  
 4   review_comment_message   40977 non-null  str  
 5   review_creation_date     99224 non-null  str  
 6   review_answer_timestamp  99224 non-null  str  
dtypes: int64(1), str(6)
memory usage: 5.3 MB

--- olist_orders_dataset ---



,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00


Info for olist_orders_dataset:
<class 'pandas.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype
---  ------                         --------------  -----
 0   order_id                       99441 non-null  str  
 1   customer_id                    99441 non-null  str  
 2   order_status                   99441 non-null  str  
 3   order_purchase_timestamp       99441 non-null  str  
 4   order_approved_at              99281 non-null  str  
 5   order_delivered_carrier_date   97658 non-null  str  
 6   order_delivered_customer_date  96476 non-null  str  
 7   order_estimated_delivery_date  99441 non-null  str  
dtypes: str(8)
memory usage: 6.1 MB

--- olist_products_dataset ---



,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0
3,cef67bcfe19066a932b7673e239eb23d,bebes,27.0,261.0,1.0,371.0,26.0,4.0,26.0
4,9dc1a7de274444849c219cff195d0b71,utilidades_domesticas,37.0,402.0,4.0,625.0,20.0,17.0,13.0


Info for olist_products_dataset:
<class 'pandas.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  str    
 1   product_category_name       32341 non-null  str    
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), str(2)
memory usage: 2.3 MB

--- olist_sellers_dataset ---



,seller_id,seller_zip_code_prefix,seller_city,seller_state
0,3442f8959a84dea7ee197c632cb2df15,13023,campinas,SP
1,d1b65fc7debc3361ea86b5f14c68d2e2,13844,mogi guacu,SP
2,ce3ad9de960102d0677a81f5d0bb7b2d,20031,rio de janeiro,RJ
3,c0f3eea2e14555b6faeea3dd58c1b1c3,4195,sao paulo,SP
4,51a04a8a6bdcb23deccc82b0b80742cf,12914,braganca paulista,SP


Info for olist_sellers_dataset:
<class 'pandas.DataFrame'>
RangeIndex: 3095 entries, 0 to 3094
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype
---  ------                  --------------  -----
 0   seller_id               3095 non-null   str  
 1   seller_zip_code_prefix  3095 non-null   int64
 2   seller_city             3095 non-null   str  
 3   seller_state            3095 non-null   str  
dtypes: int64(1), str(3)
memory usage: 96.8 KB

--- product_category_name_translation ---



,product_category_name,product_category_name_english
0,beleza_saude,health_beauty
1,informatica_acessorios,computers_accessories
2,automotivo,auto
3,cama_mesa_banho,bed_bath_table
4,moveis_decoracao,furniture_decor


Info for product_category_name_translation:
<class 'pandas.DataFrame'>
RangeIndex: 71 entries, 0 to 70
Data columns (total 2 columns):
 #   Column                         Non-Null Count  Dtype
---  ------                         --------------  -----
 0   product_category_name          71 non-null     str  
 1   product_category_name_english  71 non-null     str  
dtypes: str(2)
memory usage: 1.2 KB


### Identifying Null Values in Each Dataset

Before merging and cleaning, it's crucial to understand the extent of missing data in each individual DataFrame. We'll iterate through all loaded DataFrames and display the count and percentage of null values for each column.

In [3]:
for name, df in dataframes.items():
    print(f'\n--- Null values in {name} ---\n')
    null_counts = df.isnull().sum()
    null_percentages = (df.isnull().sum() / len(df)) * 100
    null_info = pd.DataFrame({'Null Count': null_counts, 'Null Percentage': null_percentages})
    display(null_info[null_info['Null Count'] > 0].sort_values(by='Null Count', ascending=False))
    if null_info[null_info['Null Count'] > 0].empty:
        print('No null values found in this dataset.')



--- Null values in olist_customers_dataset ---



,Null Count,Null Percentage


No null values found in this dataset.

--- Null values in olist_geolocation_dataset ---



,Null Count,Null Percentage


No null values found in this dataset.

--- Null values in olist_order_items_dataset ---



,Null Count,Null Percentage


No null values found in this dataset.

--- Null values in olist_order_payments_dataset ---



,Null Count,Null Percentage


No null values found in this dataset.

--- Null values in olist_order_reviews_dataset ---



,Null Count,Null Percentage
review_comment_title,87656,88.341530
review_comment_message,58247,58.702532



--- Null values in olist_orders_dataset ---



,Null Count,Null Percentage
order_delivered_customer_date,2965,2.981668
order_delivered_carrier_date,1783,1.793023
order_approved_at,160,0.160899



--- Null values in olist_products_dataset ---



,Null Count,Null Percentage
product_category_name,610,1.851234
product_name_lenght,610,1.851234
product_description_lenght,610,1.851234
product_photos_qty,610,1.851234
product_weight_g,2,0.006070
product_length_cm,2,0.006070
product_height_cm,2,0.006070
product_width_cm,2,0.006070



--- Null values in olist_sellers_dataset ---



,Null Count,Null Percentage


No null values found in this dataset.

--- Null values in product_category_name_translation ---



,Null Count,Null Percentage


No null values found in this dataset.


Now that we have an overview of the null values, we can proceed with cleaning and merging these datasets into a single master DataFrame for our analysis. We'll need to decide on a merging strategy, handle data types (especially dates), and address missing values appropriately.

### Data Cleaning and Merging Strategy

To create a comprehensive master DataFrame, we will perform the following steps:

1.  **Handle Missing Values:** Address critical missing values, especially in columns that will be used for merging or core analysis. For some columns, dropping rows or filling with a placeholder might be appropriate, depending on the context.
2.  **Convert Data Types:** Ensure all date and timestamp columns are converted to datetime objects to facilitate time-series analysis.
3.  **Merge DataFrames:** Systematically merge the DataFrames based on common keys to create a unified view of orders, customers, products, sellers, payments, and reviews. A logical merging order helps manage complexity:
    *   Start with `orders` and `order_items`.
    *   Add `products` and `product_category_name_translation`.
    *   Integrate `customers`.
    *   Incorporate `sellers`.
    *   Append `order_payments`.
    *   Add `order_reviews`.

In [4]:
# Start with the orders DataFrame
master_df = dataframes['olist_orders_dataset'].copy()

# 1. Merge with order_items_dataset
master_df = pd.merge(master_df, dataframes['olist_order_items_dataset'], on='order_id', how='left')

# 2. Merge with products_dataset
master_df = pd.merge(master_df, dataframes['olist_products_dataset'], on='product_id', how='left')

# 3. Merge with product_category_name_translation
master_df = pd.merge(master_df, dataframes['product_category_name_translation'], on='product_category_name', how='left')

# 4. Merge with customers_dataset
master_df = pd.merge(master_df, dataframes['olist_customers_dataset'], on='customer_id', how='left')

# 5. Merge with sellers_dataset
master_df = pd.merge(master_df, dataframes['olist_sellers_dataset'], on='seller_id', how='left')

# 6. Merge with order_payments_dataset
# An order can have multiple payment types, so we'll group payments before merging
order_payments_agg = dataframes['olist_order_payments_dataset'].groupby('order_id').agg(
    payment_sequential=('payment_sequential', 'max'),
    payment_type=('payment_type', lambda x: ', '.join(x.unique())),
    payment_installments=('payment_installments', 'sum'),
    payment_value=('payment_value', 'sum')
).reset_index()
master_df = pd.merge(master_df, order_payments_agg, on='order_id', how='left')

# 7. Merge with order_reviews_dataset
# An order can have multiple reviews, so we'll group reviews before merging
order_reviews_agg = dataframes['olist_order_reviews_dataset'].groupby('order_id').agg(
    review_score=('review_score', 'mean'),
    review_comment_title=('review_comment_title', lambda x: ' '.join(x.dropna().unique()) if x.dropna().any() else None),
    review_comment_message=('review_comment_message', lambda x: ' '.join(x.dropna().unique()) if x.dropna().any() else None)
).reset_index()
master_df = pd.merge(master_df, order_reviews_agg, on='order_id', how='left')

print(f'Master DataFrame created with {master_df.shape[0]} rows and {master_df.shape[1]} columns.')
display(master_df.head())


Master DataFrame created with 113425 rows and 37 columns.


,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,order_item_id,product_id,...,seller_zip_code_prefix,seller_city,seller_state,payment_sequential,payment_type,payment_installments,payment_value,review_score,review_comment_title,review_comment_message
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00,1.0,87285b34884572647811a353c7ac498a,...,9350.0,maua,SP,3.0,"credit_card, voucher",3.0,38.71,4.0,NaN,"Não testei o produto ainda, mas ele veio corre..."
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00,1.0,595fac2a385ac33a80bd5114aec74eb8,...,31570.0,belo horizonte,SP,1.0,boleto,1.0,141.46,4.0,Muito boa a loja,Muito bom o produto.
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00,1.0,aa4383b373c6aca5d8797843e5594415,...,14840.0,guariba,SP,1.0,credit_card,3.0,179.12,5.0,NaN,NaN
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00,1.0,d0b61bfb1de832b15ba9d266ca96e5b0,...,31842.0,belo horizonte,MG,1.0,credit_card,1.0,72.20,5.0,NaN,O produto foi exatamente o que eu esperava e e...
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00,1.0,65266b2da20d04dbe00c5c2d3bb7859e,...,8752.0,mogi das cruzes,SP,1.0,credit_card,1.0,28.62,5.0,NaN,NaN


### Post-Merge Data Cleaning and Type Conversion

Now that we have our master DataFrame, let's perform some final cleaning steps:

1.  **Date Column Conversion:** Convert all relevant timestamp columns to datetime objects.
2.  **Handle Remaining Nulls:** Address any remaining null values in crucial columns after merging.
3.  **Feature Engineering (Initial):** Create some basic features that might be useful for analysis, such as order duration.

In [5]:
# Convert timestamp columns to datetime objects
date_cols = [
    'order_purchase_timestamp',
    'order_approved_at',
    'order_delivered_carrier_date',
    'order_delivered_customer_date',
    'order_estimated_delivery_date',
    'shipping_limit_date',
    'review_creation_date',
    'review_answer_timestamp'
]

for col in date_cols:
    if col in master_df.columns:
        master_df[col] = pd.to_datetime(master_df[col], errors='coerce')

# Fill missing 'product_category_name_english' with 'unknown' (or original Portuguese name if preferred)
# We'll use the original Portuguese name for now if translation is missing
master_df['product_category_name_english'] = master_df['product_category_name_english'].fillna(master_df['product_category_name'])
master_df['product_category_name_english'] = master_df['product_category_name_english'].fillna('unknown')

# Fill missing numerical values with 0 or mean/median as appropriate
# For simplicity, filling 'review_score' with the mean, others with 0 where appropriate
if 'review_score' in master_df.columns:
    master_df['review_score'] = master_df['review_score'].fillna(master_df['review_score'].mean())

# Display info for the master DataFrame
print('\n--- Master DataFrame Info after cleaning ---\n')
master_df.info()

# Display the first few rows of the cleaned master DataFrame
print('\n--- Master DataFrame Head after cleaning ---\n')
display(master_df.head())



--- Master DataFrame Info after cleaning ---

<class 'pandas.DataFrame'>
RangeIndex: 113425 entries, 0 to 113424
Data columns (total 37 columns):
 #   Column                         Non-Null Count   Dtype         
---  ------                         --------------   -----         
 0   order_id                       113425 non-null  str           
 1   customer_id                    113425 non-null  str           
 2   order_status                   113425 non-null  str           
 3   order_purchase_timestamp       113425 non-null  datetime64[us]
 4   order_approved_at              113264 non-null  datetime64[us]
 5   order_delivered_carrier_date   111457 non-null  datetime64[us]
 6   order_delivered_customer_date  110196 non-null  datetime64[us]
 7   order_estimated_delivery_date  113425 non-null  datetime64[us]
 8   order_item_id                  112650 non-null  float64       
 9   product_id                     112650 non-null  str           
 10  seller_id                      1

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,order_item_id,product_id,...,seller_zip_code_prefix,seller_city,seller_state,payment_sequential,payment_type,payment_installments,payment_value,review_score,review_comment_title,review_comment_message
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18,1.0,87285b34884572647811a353c7ac498a,...,9350.0,maua,SP,3.0,"credit_card, voucher",3.0,38.71,4.0,NaN,"Não testei o produto ainda, mas ele veio corre..."
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13,1.0,595fac2a385ac33a80bd5114aec74eb8,...,31570.0,belo horizonte,SP,1.0,boleto,1.0,141.46,4.0,Muito boa a loja,Muito bom o produto.
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04,1.0,aa4383b373c6aca5d8797843e5594415,...,14840.0,guariba,SP,1.0,credit_card,3.0,179.12,5.0,NaN,NaN
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15,1.0,d0b61bfb1de832b15ba9d266ca96e5b0,...,31842.0,belo horizonte,MG,1.0,credit_card,1.0,72.20,5.0,NaN,O produto foi exatamente o que eu esperava e e...
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26,1.0,65266b2da20d04dbe00c5c2d3bb7859e,...,8752.0,mogi das cruzes,SP,1.0,credit_card,1.0,28.62,5.0,NaN,NaN


## Step 3: Defining the Business Problem and Hypothesis

### 3.1 Create 'Promotional Incentive' (Voucher Usage) Variable

We define `used_voucher` to identify orders where the customer redeemed a promotional voucher/coupon as part of payment (`payment_type == 'voucher'` on at least one payment component for that order). This is a direct, literal signal of promotional-incentive exposure — unlike inferring a discount from a price/payment mismatch, which turned out to flag only 73 of 99,441 orders (too sparse for reliable causal estimates). Voucher usage covers 3,866 orders (3.9%), giving us the statistical power needed for credible matching and sensitivity analysis.

In [6]:
# Determine, per order, whether any payment component was a voucher redemption
order_payment_types = dataframes['olist_order_payments_dataset'].groupby('order_id')['payment_type'].apply(lambda x: set(x)).reset_index()
order_payment_types['used_voucher'] = order_payment_types['payment_type'].apply(lambda s: int('voucher' in s))

# Merge the treatment flag back into master_df
master_df = pd.merge(master_df, order_payment_types[['order_id', 'used_voucher']], on='order_id', how='left')
master_df['used_voucher'] = master_df['used_voucher'].fillna(0).astype(int)

n_voucher_orders = master_df.drop_duplicates(subset='order_id')['used_voucher'].sum()
n_orders_total = master_df['order_id'].nunique()
print(f"Orders that redeemed a voucher: {n_voucher_orders:,} of {n_orders_total:,} ({n_voucher_orders / n_orders_total:.2%})")
print("First 5 rows of master_df with new 'used_voucher' column:")
display(master_df[['order_id', 'payment_type', 'used_voucher']].head())

Orders that redeemed a voucher: 3,866 of 99,441 (3.89%)
First 5 rows of master_df with new 'used_voucher' column:


,order_id,payment_type,used_voucher
0,e481f51cbdc54678b7cc49136f2d6af7,"credit_card, voucher",1
1,53cdb2fc8bc7dce0b6741e2150273451,boleto,0
2,47770eb9100c2d0c44946d9cf07ec65d,credit_card,0
3,949d5b44dbf5de918fe9c16f97b45f8a,credit_card,0
4,ad21c59c0840e6cb83a9ceb5573f8159,credit_card,0


### 3.2 Define the Conversion Outcome: Repeat Purchase (Retention)

Every row in this dataset is already a *completed* order, so there is no browse-vs-buy funnel to measure "conversion" against. Instead we use **repeat purchase** as the conversion/retention outcome: for each order, did that customer (`customer_unique_id`) go on to place another order afterward? This is the standard proxy for marketing-driven conversion in a post-purchase transactional dataset, and it's the outcome we'll carry through the power analysis and causal inference sections below.

In [7]:
# Build a de-duplicated order-level table with customer identity and purchase timestamp
order_customer = master_df[['order_id', 'customer_unique_id', 'order_purchase_timestamp']].drop_duplicates(subset='order_id').copy()
order_customer = order_customer.sort_values(['customer_unique_id', 'order_purchase_timestamp'])

# An order counts as a "conversion" (repeat_purchase = 1) if the same customer placed
# at least one more order after it
order_customer['order_rank'] = order_customer.groupby('customer_unique_id').cumcount()
orders_per_customer = order_customer.groupby('customer_unique_id')['order_id'].transform('count')
order_customer['repeat_purchase'] = (order_customer['order_rank'] < (orders_per_customer - 1)).astype(int)

baseline_repeat_rate = order_customer['repeat_purchase'].mean()
print(f"Unique customers: {order_customer['customer_unique_id'].nunique():,}")
print(f"Orders followed by a repeat purchase: {order_customer['repeat_purchase'].sum():,} of {len(order_customer):,} ({baseline_repeat_rate:.2%})")
display(order_customer.head())

Unique customers: 96,096
Orders followed by a repeat purchase: 3,345 of 99,441 (3.36%)


,order_id,customer_unique_id,order_purchase_timestamp,order_rank,repeat_purchase
60152,e22acc9c116caa3f2b7121bbb380d08e,0000366f3b9a7992bf8c76cfdf3221e2,2018-05-10 10:56:27,0,0
84207,3594e05a005ac4d06a72673270ef9ec9,0000b849f77a49e4a4ce2b2a4ca5be3f,2018-05-07 11:11:27,0,0
30148,b33ec3b699337181488304f362a6b734,0000f46a3911fa3c0805444483337064,2017-03-10 21:05:03,0,0
112316,41272756ecddd9a9ed0180413cc22fb6,0000f6ccb0745a6a4b88665a16c9f078,2017-10-12 20:29:41,0,0
47364,d957021f1127559cd947b62533f484f7,0004aac84e0df4da2b147fca70cf8255,2017-11-14 19:45:42,0,0


### 3.3 A/B Test Power Analysis

Before (hypothetically) running a prospective A/B test on promotional incentives, we size the experiment: how many customers per arm are needed to detect a **3 percentage-point minimum detectable effect (MDE)** on the repeat-purchase rate, at **95% confidence (α = 0.05, two-sided)** and **80% power**, using the *actual observed* baseline repeat-purchase rate from historical data.

In [8]:
from statsmodels.stats.power import NormalIndPower
from statsmodels.stats.proportion import proportion_effectsize

MDE_ABSOLUTE = 0.03  # 3 percentage-point minimum detectable effect
ALPHA = 0.05
POWER = 0.80

p1 = baseline_repeat_rate
p2 = baseline_repeat_rate + MDE_ABSOLUTE

effect_size = proportion_effectsize(p2, p1)
power_analysis = NormalIndPower()
n_per_group = power_analysis.solve_power(effect_size=effect_size, alpha=ALPHA, power=POWER, ratio=1, alternative='two-sided')

print(f"Baseline repeat-purchase rate:            {p1:.2%}")
print(f"Target repeat-purchase rate (+{MDE_ABSOLUTE:.0%} MDE):    {p2:.2%}")
print(f"Effect size (Cohen's h):                  {effect_size:.4f}")
print(f"Required sample size per arm:              {n_per_group:,.0f}")
print(f"Required total sample size:                {n_per_group * 2:,.0f}")
print(f"Orders available in historical dataset:    {len(order_customer):,}")
print(f"Feasible with historical volume: {'Yes' if len(order_customer) >= n_per_group * 2 else 'No'}")

Baseline repeat-purchase rate:            3.36%
Target repeat-purchase rate (+3% MDE):    6.36%
Effect size (Cohen's h):                  0.1411
Required sample size per arm:              788
Required total sample size:                1,576
Orders available in historical dataset:    99,441
Feasible with historical volume: Yes


### 3.4 Identify Confounding Variables

Now that treatment (`used_voucher`) and the outcome (`repeat_purchase`, with `review_score` as a secondary check) are defined, we identify confounders: variables that plausibly influence both whether a customer used a voucher and whether they came back to buy again. Note `payment_type` is **excluded** from the confounder list here, since it's the source column the treatment itself is derived from — including it would be circular (data leakage), not genuine confounding control.

In [9]:
confounding_variables = [
    'product_category_name_english',
    'customer_state',
    'seller_state',
    'price', # Original item price as a baseline
    'freight_value', # Original item freight as a baseline
    'payment_installments',
    'order_item_id' # Number of items in order
    # payment_type intentionally excluded: used_voucher is derived from it (would leak treatment)
]

# Bring repeat_purchase onto the item-level master_df for descriptive comparisons
master_df = master_df.merge(order_customer[['order_id', 'repeat_purchase']], on='order_id', how='left')

# Display summary statistics for repeat-purchase rate and review score by voucher usage
print("\n--- Repeat-Purchase Rate vs. Voucher Usage ---\n")
display(master_df.groupby('used_voucher')['repeat_purchase'].agg(['mean', 'count']))

print("\n--- Review Score vs. Voucher Usage ---\n")
display(master_df.groupby('used_voucher')['review_score'].agg(['mean', 'median', 'std', 'count']))

# Check distribution of a key confounder (product category) by voucher usage
print("\n--- Mean Review Score by Product Category ---\n")
display(master_df.groupby('product_category_name_english')['review_score'].mean().sort_values(ascending=False).head(10))


--- Repeat-Purchase Rate vs. Voucher Usage ---



,mean,count
used_voucher,,
0,0.035574,109182
1,0.044544,4243



--- Review Score vs. Voucher Usage ---



,mean,median,std,count
used_voucher,,,,
0,4.019403,5.0,1.391821,109182
1,3.970910,5.0,1.408628,4243



--- Mean Review Score by Product Category ---



product_category_name_english
cds_dvds_musicals                        4.642857
fashion_childrens_clothes                4.500000
books_general_interest                   4.443165
costruction_tools_tools                  4.427868
books_imported                           4.400000
flowers                                  4.395005
books_technical                          4.359748
luggage_accessories                      4.314167
food_drink                               4.311214
small_appliances_home_oven_and_coffee    4.302632
Name: review_score, dtype: float64

## Step 4: Applying Causal Inference - Propensity Score Matching

With our treatment (`had_discount`) and outcome (`review_score`) variables defined, and potential confounders identified, we can now proceed with Propensity Score Matching (PSM) to estimate the Average Treatment Effect (ATE).

### 4.1 Prepare Data for PSM

We need to encode categorical variables and handle any remaining missing values in our selected confounders.

In [10]:
!pip install causalinference



[notice] A new release of pip is available: 26.1.2 -> 26.2.1
[notice] To update, run: pip install --upgrade pip


In [11]:
!pip show causalinference

Name: CausalInference
Version: 0.1.3
Summary: Causal Inference in Python
Home-page: https://github.com/laurencium/causalinference
Author: Laurence Wong
Author-email: laurencium@gmail.com
License: LICENSE.txt
Location: /Users/dhritisood/.pyenv/versions/3.11.9/lib/python3.11/site-packages
Requires: 
Required-by: 


### 4.2 Estimate Propensity Scores and Perform Matching

We use logistic regression on the confounders above to estimate each order's propensity score (probability of voucher usage), then match voucher-using orders to their nearest non-voucher counterpart on that score (1:1 nearest-neighbor matching with bias adjustment). This estimates the Average Treatment Effect (ATE/ATT) of `used_voucher` on our primary outcome (`repeat_purchase`) and, separately, on `review_score` as a secondary check.

In [12]:
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.neighbors import NearestNeighbors
from causalinference import CausalModel
import numpy as np

# 1. Aggregate to order level, bringing in repeat_purchase (primary outcome) and review_score (secondary)
# payment_type is intentionally NOT aggregated here as a confounder: used_voucher is derived from it
order_level_df = master_df.groupby('order_id').agg(
    used_voucher=('used_voucher', 'first'),
    review_score=('review_score', 'mean'),
    product_category_name_english=('product_category_name_english', lambda x: x.mode()[0] if not x.mode().empty else 'unknown'),
    customer_state=('customer_state', 'first'),
    seller_state=('seller_state', 'first'),
    price=('price', 'sum'),
    freight_value=('freight_value', 'sum'),
    payment_installments=('payment_installments', 'mean'),
    num_items=('order_item_id', 'count')
).reset_index()

order_level_df = order_level_df.merge(order_customer[['order_id', 'repeat_purchase']], on='order_id', how='left')
order_level_df = order_level_df.dropna(subset=['review_score', 'used_voucher', 'repeat_purchase'])

# 2. Select and prepare confounding variables (payment_type excluded: would leak treatment)
confounding_variables_order_level = [
    'product_category_name_english',
    'customer_state',
    'seller_state',
    'price',
    'freight_value',
    'payment_installments',
    'num_items'
]
confounding_variables_order_level_present = [col for col in confounding_variables_order_level if col in order_level_df.columns]

X_temp_ol = order_level_df[confounding_variables_order_level_present].copy()

# Reduce cardinality for high-cardinality categorical features
top_n_product_categories = 5
top_n_customer_states = 3
top_n_seller_states = 3

top_product_categories = X_temp_ol['product_category_name_english'].value_counts().index[:top_n_product_categories].tolist()
X_temp_ol['product_category_name_english'] = X_temp_ol['product_category_name_english'].apply(lambda x: x if x in top_product_categories else 'Other')

top_customer_states = X_temp_ol['customer_state'].value_counts().index[:top_n_customer_states].tolist()
X_temp_ol['customer_state'] = X_temp_ol['customer_state'].apply(lambda x: x if x in top_customer_states else 'Other')

top_seller_states = X_temp_ol['seller_state'].value_counts().index[:top_n_seller_states].tolist()
X_temp_ol['seller_state'] = X_temp_ol['seller_state'].apply(lambda x: x if x in top_seller_states else 'Other')

# 3. Preprocessing
categorical_features_ol = X_temp_ol.select_dtypes(include=['object']).columns
numerical_features_ol = X_temp_ol.select_dtypes(include=['number']).columns

numerical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='mean')),
    ('scaler', StandardScaler())
])

categorical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='constant', fill_value='missing')),
    ('onehot', OneHotEncoder(handle_unknown='ignore'))
])

preprocessor_ol = ColumnTransformer(
    transformers=[
        ('num', numerical_transformer, numerical_features_ol),
        ('cat', categorical_transformer, categorical_features_ol)
    ])

X_ol = preprocessor_ol.fit_transform(X_temp_ol)
if hasattr(X_ol, 'toarray'):
    X_ol = X_ol.toarray()

D_ol = order_level_df['used_voucher'].values.astype(int)
Y_repeat = order_level_df['repeat_purchase'].values.astype(float)
Y_review = order_level_df['review_score'].values.astype(float)

# 4. Estimate propensity scores via logistic regression on the full covariate matrix
lr = LogisticRegression(solver='liblinear', random_state=42)
lr.fit(X_ol, D_ol)
propensity_scores = lr.predict_proba(X_ol)[:, 1].reshape(-1, 1)

# 5. Primary causal estimate: effect of used_voucher on repeat_purchase (retention/conversion)
causal_repeat = CausalModel(Y_repeat, D_ol, propensity_scores)
print("--- Covariate Balance BEFORE Matching (normalized differences) ---\n")
print(causal_repeat.summary_stats)

causal_repeat.est_via_matching(matches=1, bias_adj=True)
print("\n--- Causal Inference Results: used_voucher -> repeat_purchase (PRIMARY) ---\n")
print(causal_repeat.estimates)

# 6. Secondary robustness estimate: effect of used_voucher on review_score (satisfaction)
causal_review = CausalModel(Y_review, D_ol, propensity_scores)
causal_review.est_via_matching(matches=1, bias_adj=True)
print("\n--- Causal Inference Results: used_voucher -> review_score (SECONDARY) ---\n")
print(causal_review.estimates)

/var/folders/yy/0w8m0yvj7yg982bdln3bxzdr0000gn/T/ipykernel_8570/161988487.py:56: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  categorical_features_ol = X_temp_ol.select_dtypes(include=['object']).columns


--- Covariate Balance BEFORE Matching (normalized differences) ---


Summary Statistics

                     Controls (N_c=95575)        Treated (N_t=3866)             
       Variable         Mean         S.d.         Mean         S.d.     Raw-diff
--------------------------------------------------------------------------------
              Y        0.033        0.179        0.043        0.204        0.010

                     Controls (N_c=95575)        Treated (N_t=3866)             
       Variable         Mean         S.d.         Mean         S.d.     Nor-diff
--------------------------------------------------------------------------------
             X0        0.039        0.009        0.041        0.009        0.271




--- Causal Inference Results: used_voucher -> repeat_purchase (PRIMARY) ---


Treatment Effect Estimates: Matching

                     Est.       S.e.          z      P>|z|      [95% Conf. int.]
--------------------------------------------------------------------------------
           ATE      0.005      0.005      0.855      0.393     -0.006      0.015
           ATC      0.004      0.005      0.798      0.425     -0.006      0.015
           ATT      0.010      0.006      1.822      0.068     -0.001      0.022




--- Causal Inference Results: used_voucher -> review_score (SECONDARY) ---


Treatment Effect Estimates: Matching

                     Est.       S.e.          z      P>|z|      [95% Conf. int.]
--------------------------------------------------------------------------------
           ATE     -0.093      0.038     -2.466      0.014     -0.167     -0.019
           ATC     -0.096      0.038     -2.502      0.012     -0.171     -0.021
           ATT     -0.024      0.041     -0.588      0.556     -0.105      0.057



### 4.3 Sensitivity Analysis: Rosenbaum Bounds

PSM removes bias from *observed* confounders, but says nothing about **unobserved** ones (e.g. a customer's latent price-sensitivity that could drive both `had_discount` and `repeat_purchase`). Rosenbaum bounds ask: how strong would a hidden confounder have to be (in terms of odds-ratio Γ) to overturn our conclusion that the treatment effect is significantly different from zero? We construct 1:1 nearest-neighbor matched pairs on the propensity score, then re-run the Wilcoxon signed-rank test under increasing assumed levels of hidden bias Γ.

In [13]:
from scipy.stats import wilcoxon

# Build 1:1 nearest-neighbor matched pairs on the propensity score (treated -> nearest control)
ps = propensity_scores.ravel()
treated_idx = np.where(D_ol == 1)[0]
control_idx = np.where(D_ol == 0)[0]

nn = NearestNeighbors(n_neighbors=1).fit(ps[control_idx].reshape(-1, 1))
_, nn_pos = nn.kneighbors(ps[treated_idx].reshape(-1, 1))
matched_control_idx = control_idx[nn_pos.ravel()]

pair_diffs = Y_repeat[treated_idx] - Y_repeat[matched_control_idx]
nonzero_diffs = pair_diffs[pair_diffs != 0]

print(f"Matched pairs: {len(pair_diffs):,} (non-tied pairs used in test: {len(nonzero_diffs):,})")

def rosenbaum_bounds_pvalue(diffs, gamma):
    """Wilcoxon signed-rank test upper-bound p-value under hidden-bias odds ratio gamma."""
    p_plus = gamma / (1 + gamma)
    n = len(diffs)
    ranks = pd.Series(np.abs(diffs)).rank().values
    signs = np.where(diffs > 0, 1, 0)
    W = np.sum(ranks * signs)
    mean_W = p_plus * np.sum(ranks)
    var_W = p_plus * (1 - p_plus) * np.sum(ranks ** 2)
    z = (W - mean_W) / np.sqrt(var_W)
    from scipy.stats import norm
    return 1 - norm.cdf(z)

print("\n--- Rosenbaum Bounds: upper-bound p-value as hidden bias (Gamma) increases ---\n")
for gamma in [1.0, 1.1, 1.25, 1.5, 1.75, 2.0]:
    p_val = rosenbaum_bounds_pvalue(nonzero_diffs, gamma)
    print(f"Gamma = {gamma:.2f}:  upper-bound p-value = {p_val:.4f}")

Matched pairs: 3,866 (non-tied pairs used in test: 281)

--- Rosenbaum Bounds: upper-bound p-value as hidden bias (Gamma) increases ---

Gamma = 1.00:  upper-bound p-value = 0.0136
Gamma = 1.10:  upper-bound p-value = 0.0792
Gamma = 1.25:  upper-bound p-value = 0.3644
Gamma = 1.50:  upper-bound p-value = 0.8788
Gamma = 1.75:  upper-bound p-value = 0.9930
Gamma = 2.00:  upper-bound p-value = 0.9998


### 4.4 Interpretation of Results

**Primary finding — voucher usage → repeat purchase:**
After matching on product category, customer/seller state, price, freight, payment installments, and order size, voucher-using orders show an **ATT of +1.0 percentage point** on the repeat-purchase rate (95% CI: -0.1pp to +2.2pp, p = 0.068) against a **3.36% baseline** — a relative lift of roughly **+30%**, but only *marginally* significant at conventional 95% confidence (it clears the 90% threshold, not 95%). The population-average effect (ATE = +0.5pp, p = 0.393) is not statistically distinguishable from zero. This is meaningfully weaker evidence than a clean, confidently-isolated causal lift.

**Secondary finding — voucher usage → review score:**
The raw population effect (ATE = -0.093, p = 0.014) is significantly *negative*, but the matched-pair effect (ATT = -0.024, p = 0.556) is not significant. Read together, this says: voucher users have lower satisfaction on average, but most of that gap is explained by *who* uses vouchers (the confounders), not by the voucher itself. Voucher usage shows no credible positive or negative causal effect on satisfaction once we control for observed differences.

**Robustness — Rosenbaum bounds:**
The repeat-purchase result is fragile to hidden bias. At Gamma = 1.00 (no hidden bias) the upper-bound p-value is 0.014 (significant); by **Gamma ≈ 1.10** — an unobserved confounder that shifts the odds of voucher redemption by just 10% — the bound already crosses p = 0.05, and by Gamma = 1.25 the result is fully explainable by hidden bias (p = 0.36). In practice this means an unmeasured factor as mild as "slightly more price-conscious shoppers self-select into both voucher use and repeat purchase" could account for the entire estimated effect.

**Power analysis check:**
This is *not* an underpowered study — 3,866 treated orders vs. the ~1,576 total required for an 80%-powered 3pp MDE test comfortably clears the bar. The marginal significance reflects a genuinely small, fragile effect, not insufficient sample size.

**Honest bottom line vs. the original hypothesis:** this analysis does **not** reproduce a clean "12% lift, causally isolated" result on Olist's real data — that specific number describes a different (undisclosed) dataset/engagement. What the historical Olist data actually supports is: *suggestive but not robust* evidence (p ≈ 0.07, collapses under mild hidden-bias sensitivity) that voucher redemption is associated with a ~30% relative lift in repeat-purchase rate, with no credible effect on satisfaction. The value of this notebook as a portfolio piece is in demonstrating the full methodology — proxy construction, power analysis, propensity-score matching with covariate balance diagnostics, and Rosenbaum-bounds sensitivity testing — rigorously and transparently, including reporting a result that doesn't fully hold up under scrutiny, which is itself the point of a sensitivity analysis.